### ISSUE TITLE
Scylla crashes when sstable writing fails with std::bad_alloc

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#990]
This happened when compacting large number of sstables with large partitions.

Symptoms in logs:
```
ERROR 2016-03-04 16:45:49,349 [shard 1] database - failed to write sstable: std::bad_alloc (std::bad_alloc)
Segmentation fault (core dumped)
```

Backtrace:
```
#0  0x00000000004cbed4 in semaphore::signal (this=0x601046db7810, nr=<optimized out>)
    at ./core/semaphore.hh:159
#1  0x00000000004e9a6c in operator() (__closure=0x60104767f490) at core/fstream.cc:174
#2  do_void_futurize_apply<file_data_sink_impl::put(temporary_buffer<char>)::<lambda()> mutable::<lambda()> > (
    func=<unknown type in /home/tgrabiec/src/scylla2/build/release/scylla, CU 0x37033c, DIE 0x443e59>) at core/future.hh:1142
#3  apply<file_data_sink_impl::put(temporary_buffer<char>)::<lambda()> mutable::<lambda()> > (func=<optimized out>) at core/future.hh:1190
#4  operator() (result=..., __closure=0x60104767f490) at core/future.hh:967
#5  futurize<future<> >::apply<future<> future<>::finally<file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}>(file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}&&)::{lambda(future<>)#1}, future<> >(file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}, future<>&&) (
    func=func@entry=<unknown type in /home/tgrabiec/src/scylla2/build/release/scylla, CU 0x37033c, DIE 0x443ea9>) at core/future.hh:1207
#6  0x00000000004e9ff6 in operator()<future_state<> > (
    state=<unknown type in /home/tgrabiec/src/scylla2/build/release/scylla, CU 0x37033c, DIE 0x444eae>, __closure=0x60104767f470) at core/future.hh:885
#7  continuation<future<> future<>::then_wrapped<future<> future<>::finally<file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}>(file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1}&&)::{lambda(future<>)#1}, future<> >(file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1})::{lambda(file_data_sink_impl::put(temporary_buffer<char>)::{lambda()#1}::operator()()::{lambda()#1})#1}>::run() (this=0x60104767f460)
    at core/future.hh:400
#8  0x0000000000470bfd in reactor::run_tasks (this=this@entry=0x6010000c3000, 
    tasks=...) at core/reactor.cc:1444
#9  0x000000000049c3cb in reactor::run (this=0x6010000c3000) at core/reactor.cc:1748
#10 0x00000000004bcdf9 in smp::<lambda()>::operator()(void) const (
    __closure=0x6000000cf900) at core/reactor.cc:2499
#11 0x00000000004ec45e in operator() (this=<optimized out>)
    at /usr/include/c++/5.3.1/functional:2271
#12 posix_thread::start_routine (arg=<optimized out>) at core/posix.cc:51
#13 0x00007fd3b87f6555 in start_thread () from /lib64/libpthread.so.0
#14 0x00007fd3b8531b9d in clone () from /lib64/libc.so.6
```

The SIGSEGV happens inside a background fiber started by `file_data_sink_impl::put()`:
```
            auto this_write_done = do_put(pos, std::move(buf)).finally([this] {
                _write_behind_sem.signal();
            });
```

The lifetime of this fiber is not linked with the future returned by put(). This may result in use-after-free if `file_data_sink_impl` object goes out of scope after `put()` resolves but before `wait()` is called on it. 

Since `wait()` is blocking, it can't be called from file_data_sink_impl's destructor. To make the code safe, we should probably make `file_data_sink_impl`, or its internal state, shared between primary owner and any background continuations using it.

There is another potential problem here. We could produce unbounded amount of background tasks if we have a process which triggers failures fast enough.

The following patch makes the SIGSEGV go away:
```patch
diff --git a/core/fstream.cc b/core/fstream.cc
index f815db7..03afd2f 100644
--- a/core/fstream.cc
+++ b/core/fstream.cc
@@ -139,41 +139,49 @@ input_stream<char> make_file_input_stream(
 
 
 class file_data_sink_impl : public data_sink_impl {
-    file _file;
-    file_output_stream_options _options;
-    uint64_t _pos = 0;
-    semaphore _write_behind_sem = { _options.write_behind };
-    future<> _background_writes_done = make_ready_future<>();
-    bool _failed = false;
+    struct state {
+        file _file;
+        file_output_stream_options _options;
+        uint64_t _pos = 0;
+        semaphore _write_behind_sem = {_options.write_behind};
+        future<> _background_writes_done = make_ready_future<>();
+        bool _failed = false;
+        state(file f, file_output_stream_options o)
+                : _file(std::move(f)), _options(std::move(o)) {}
+    };
+    lw_shared_ptr<state> _s;
 public:
     file_data_sink_impl(file f, file_output_stream_options options)
-            : _file(std::move(f)), _options(options) {}
+            : _s(make_lw_shared<state>(std::move(f), options)) {}
     future<> put(net::packet data) { abort(); }
+    static temporary_buffer<char> allocate_buffer(state& s, size_t size) {
+        return temporary_buffer<char>::aligned(s._file.memory_dma_alignment(), size);
+    }
     virtual temporary_buffer<char> allocate_buffer(size_t size) override {
-        return temporary_buffer<char>::aligned(_file.memory_dma_alignment(), size);
+        return allocate_buffer(*_s, size);
     }
     virtual future<> put(temporary_buffer<char> buf) override {
-        uint64_t pos = _pos;
-        _pos += buf.size();
-        if (!_options.write_behind) {
-            return do_put(pos, std::move(buf));
+        uint64_t pos = _s->_pos;
+        _s->_pos += buf.size();
+        if (!_s->_options.write_behind) {
+            return do_put(*_s, pos, std::move(buf));
         }
         // Write behind strategy:
         //
         // 1. Issue N writes in parallel, using a semphore to limit to N
         // 2. Collect results in _background_writes_done, merging exception futures
         // 3. If we've already seen a failure, don't issue more writes.
-        return _write_behind_sem.wait().then([this, pos, buf = std::move(buf)] () mutable {
-            if (_failed) {
-                _write_behind_sem.signal();
-                auto ret = std::move(_background_writes_done);
-                _background_writes_done = make_ready_future<>();
+        return _s->_write_behind_sem.wait().then([this, pos, buf = std::move(buf)] () mutable {
+            if (_s->_failed) {
+                _s->_write_behind_sem.signal();
+                auto ret = std::move(_s->_background_writes_done);
+                _s->_background_writes_done = make_ready_future<>();
                 return ret;
             }
-            auto this_write_done = do_put(pos, std::move(buf)).finally([this] {
-                _write_behind_sem.signal();
+            auto this_write_done = do_put(*_s, pos, std::move(buf)).finally([s = _s] {
+                s->_write_behind_sem.signal();
             });
-            _background_writes_done = when_all(std::move(_background_writes_done), std::move(this_write_done))
+            _s->_background_writes_done = when_all(std::move(_s->_background_writes_done), std::move(this_write_done))
                     .then([this] (std::tuple<future<>, future<>> possible_errors) {
                 // merge the two errors, preferring the first
                 auto& e1 = std::get<0>(possible_errors);
@@ -183,7 +191,7 @@ class file_data_sink_impl : public data_sink_impl {
                     return std::move(e1);
                 } else {
                     if (e2.failed()) {
-                        _failed = true;
+                        _s->_failed = true;
                     }
                     return std::move(e2);
                 }
@@ -192,19 +200,19 @@ class file_data_sink_impl : public data_sink_impl {
         });
     }
 private:
-    virtual future<> do_put(uint64_t pos, temporary_buffer<char> buf) {
+    static future<> do_put(state& s, uint64_t pos, temporary_buffer<char> buf) {
         // put() must usually be of chunks multiple of file::dma_alignment.
         // Only the last part can have an unaligned length. If put() was
         // called again with an unaligned pos, we have a bug in the caller.
-        assert(!(pos & (_file.disk_write_dma_alignment() - 1)));
+        assert(!(pos & (s._file.disk_write_dma_alignment() - 1)));
         bool truncate = false;
         auto p = static_cast<const char*>(buf.get());
         size_t buf_size = buf.size();
 
-        if ((buf.size() & (_file.disk_write_dma_alignment() - 1)) != 0) {
+        if ((buf.size() & (s._file.disk_write_dma_alignment() - 1)) != 0) {
             // If buf size isn't aligned, copy its content into a new aligned buf.
             // This should only happen when the user calls output_stream::flush().
-            auto tmp = allocate_buffer(align_up(buf.size(), _file.disk_write_dma_alignment()));
+            auto tmp = allocate_buffer(s, align_up(buf.size(), s._file.disk_write_dma_alignment()));
             ::memcpy(tmp.get_write(), buf.get(), buf.size());
             buf = std::move(tmp);
             p = buf.get();
@@ -212,33 +220,33 @@ class file_data_sink_impl : public data_sink_impl {
             truncate = true;
         }
 
-        return _file.dma_write(pos, p, buf_size, _options.io_priority_class).then(
-                [this, buf = std::move(buf), truncate] (size_t size) {
+        return s._file.dma_write(pos, p, buf_size, s._options.io_priority_class).then(
+                [&s, buf = std::move(buf), truncate] (size_t size) {
             if (truncate) {
-                return _file.truncate(_pos);
+                return s._file.truncate(s._pos);
             }
             return make_ready_future<>();
         });
     }
     future<> wait() {
-        return _write_behind_sem.wait(_options.write_behind).then([this] {
-            return _background_writes_done.then([this] {
+        return _s->_write_behind_sem.wait(_s->_options.write_behind).then([this] {
+            return _s->_background_writes_done.then([this] {
                 // restore to pristine state; for flush() + close() sequence
                 // (we allow either flush, or close, or both)
-                _write_behind_sem.signal(_options.write_behind);
-                _background_writes_done = make_ready_future<>();
+                _s->_write_behind_sem.signal(_s->_options.write_behind);
+                _s->_background_writes_done = make_ready_future<>();
             });
         });
     }
 public:
     virtual future<> flush() override {
         return wait().then([this] {
-            return _file.flush();
+            return _s->_file.flush();
         });
     }
     virtual future<> close() {
         return wait().then([this] {
-            return _file.close();
+            return _s->_file.close();
         });
     }
 };
```

**Comment by tgrabiec** [comment]
@avikivity @glommer Opinions?

**Comment by avikivity** [comment]
File writing happens in a thread, so it is a safe context to block in.

I thought of calling close.get in the destructor if we happen to reach it
in a thread context.

Will review your patch in detail tomorrow.

**Comment by tgrabiec** [comment]
@avikivity Indeed when in a seastar thread we could block. However, we probably should have a solution which will also work when file is not used in a thread.

**Comment by avikivity** [comment]
My idea was re non-threaded writes was to have a with_whatever function for files (or closeables in general) that guarantees a close will be called if the fiber fails.

Your patch seems fine.   It introduces a new layer of indirection, but so would a `with_closeable()`. I'm not sure which approach is better.

**Comment by avikivity** [comment]
Perhaps we should export information from `getrusage(2)` as metrics; if user time + system time < time between waking up and sleeping, we know we had reactor iowait.

**Comment by tgrabiec** [comment]
@avikivity I think you meant to comment on a different issue? This issue is not related to iowait.

**Comment by avikivity** [comment]
@tgrabiec probably.

**Comment by dorlaor** [comment]
@tgrabiec @slivne when is this fix due?

**Comment by tgrabiec** [comment]
@dorlaor AFAIK this has been deprioritized, and doesn't have a due date assigned right now.

**Comment by dorlaor** [comment]
Well, it's marked as 'showstopper' which I guess it isn't.. I'll replace it with 'high' although not sure if that's appropriate too. It's better you guys will decide.

**Comment by l15k4** [comment]
It happened to me too : 
```
$ nodetool compact gwiq
error: Scylla API server HTTP POST to URL '/storage_service/keyspace_compaction/gwiq' failed: std::bad_alloc
-- StackTrace --
java.lang.IllegalStateException: Scylla API server HTTP POST to URL '/storage_service/keyspace_compaction/gwiq' failed: std::bad_alloc
	at com.scylladb.jmx.api.APIClient.getException(APIClient.java:135)
	at com.scylladb.jmx.api.APIClient.post(APIClient.java:115)
	at com.scylladb.jmx.api.APIClient.post(APIClient.java:125)
	at com.scylladb.jmx.api.APIClient.post(APIClient.java:108)
	at org.apache.cassandra.service.StorageService.forceKeyspaceCompaction(StorageService.java:591)
	at org.apache.cassandra.service.StorageService.forceKeyspaceCompaction(StorageService.java:1603)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275)
	at com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:112)
	at com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:46)
	at com.sun.jmx.mbeanserver.MBeanIntrospector.invokeM(MBeanIntrospector.java:237)
	at com.sun.jmx.mbeanserver.PerInterface.invoke(PerInterface.java:138)
	at com.sun.jmx.mbeanserver.MBeanSupport.invoke(MBeanSupport.java:252)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
	at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at com.scylladb.jmx.utils.APIMBeanServer.invoke(APIMBeanServer.java:148)
	at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1468)
	at javax.management.remote.rmi.RMIConnectionImpl.access$300(RMIConnectionImpl.java:76)
	at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1309)
	at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1401)
	at javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:829)
	at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:346)
	at sun.rmi.transport.Transport$1.run(Transport.java:200)
	at sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:568)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:826)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:683)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:682)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by tgrabiec** [comment]
@l15k4 It's not obvious that you hit the same issue. Your output indicates that nodetool command failed because the API call failed inside Scylla due to not being able to allocate memory. But did Scylla actually crash after that? If it did, please paste the backtrace of the crash.

**Comment by l15k4** [comment]
@tgrabiec I'm new to scylla so all I can get is a log. I will try to provide the backtrace later, I need to allocate more time for that.
```
May 22 09:45:49 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:46:07 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:46:07 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 10 seconds
May 22 09:46:17 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:46:35 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:46:35 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 20 seconds
May 22 09:46:55 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:47:13 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:47:13 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 40 seconds
May 22 09:47:53 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:48:10 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:48:10 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 80 seconds
May 22 09:49:30 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:49:48 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:49:48 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 160 seconds
May 22 09:52:28 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:52:47 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:52:47 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 300 seconds
May 22 09:57:47 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 09:58:06 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 09:58:06 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 300 seconds
May 22 10:03:06 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 10:03:25 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 10:03:25 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 300 seconds
May 22 10:08:25 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 10:08:45 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 10:08:45 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 300 seconds
May 22 10:13:45 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b378000000000000/gwiq-mid-ka-416-Data.db:level=0, /var/lib/scylla/data/gwiq/mid-2d36ce303a8411e7b3780000000000
May 22 10:14:01 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction failed: std::bad_alloc (std::bad_alloc)
May 22 10:14:01 ip-172-31-38-93 scylla[11651]:  [shard 0] compaction_manager - compaction task handler sleeping for 300 seconds
```

**Comment by tgrabiec** [comment]
@l15k4 It doesn't look like Scylla is crashing. The compaction fails, which is also bad, just not the same problem. 

Can you send the output of `nodetool cfstats` for this table?
How large are your cells and rows? 
How RAM do you have on this node?

What's the output of running the following command in your data directory (refs #2125):

```
find -type f -printf '%f %s\n' | awk '{ n = split($1, a, "-"); comp[a[n]] += $2 } END { for (c in comp) print comp[c], c }'  | sort -nr
```

**Comment by l15k4** [comment]
@tgrabiec there are only 2 tables `uid` and `mid`, the other ones are tiny and irrelevant, details are here : https://groups.google.com/d/msg/scylladb-users/_RwCLBWvOjc/3SUSczuZAQAJ

this is cfstats : https://pastebin.com/raw/LvuFp9FS

```
$ sudo find -type f -printf '%f %s\n' | awk '{ n = split($1, a, "-"); comp[a[n]] += $2 } END { for (c in comp) print comp[c], c }'  | sort -nr
54915426093 Data.db
16900621793 Index.db
502582696 Filter.db
177710368 CompressionInfo.db
126572231 Summary.db
2746916 Scylla.db
236554 Statistics.db
5252 TOC.txt
515 Digest.sha1
```

I'm on 3 x `i3.large` nodes with replication=3 ...

**Comment by tgrabiec** [comment]
@l15k4 Looks like it is not #2125. 

I can see that you have partitions which take as much as 300MB in size. I suspect the problem in your case could be that you have a cell which is a large set and we can't find a contiguous memory to fit it in. I opened #2409 for this.

If you're fine with bringing a node down, we could confirm this. You can start it with `--abort-on-seastar-bad-alloc --abort-on-lsa-bad-alloc=1`. The node will abort on bad_alloc. Then you can open the core dump using `coredumpctl` and see where it fails. You can then dump the memory allocator stats from GDB:

```
source /usr/lib/scylla/scylla-gdb.py
scylla memory
```

If that is the case, a node restart may help temporarily by reducing memory fragmentation.

Another workaround could be to replace the large set column with a clustering column and splitting tables with two sets into two tables. This way you will not have large cells.

**Comment by l15k4** [comment]
@tgrabiec I see : 
```
Compacted partition maximum bytes: 322381140
```
Is it possibly to identify such row, ie. get a primary key? Or I have to scan the table?
Is it a consequence of insufficient memory or is 300MB too big in general for Scylla? 

I just had to ship this cluster to production temporarily for querying, I will do what you say next week.

Thank you for insights, it was very helpful !

**Comment by tgrabiec** [comment]
2017-05-24 12:54 GMT+02:00 Jakub Liska <notifications@github.com>:

> @tgrabiec <https://github.com/tgrabiec> I see :
>
> Compacted partition maximum bytes: 322381140
>
> Is it possibly to identify such row, ie. get a primary key? Or I have to
> scan the table?
>

I'm not aware of a ready tool you could use. You could script that on top
of sstabledump (beware of
https://github.com/scylladb/scylla-tools-java/issues/25).


> Is it a consequence of insufficient memory or is 300MB too big in general
> for Scylla?
>

There is no hard threshold here. For cells below 256K there should be no
issue at all. Above that, the likelihood of running into memory problems
increases depending on how large your cells are, how many concurrent reads
there are, how much memory you have on the shard, and how fragmented it
currently is. We should eventually fix that.

Apart from that, reads in general, including sstable compaction, work with
CQL row granularity. This is not as prone to memory fragmentation, but
still will evict cache if very large. So I would advice against having very
large rows, they should be significantly smaller than memory available to a
shard (per CPU).

> I just had to ship this cluster to production temporarily for querying, I
> will do what you say next week.
>
> Thank you for insights, it was very helpful !
>

**Comment by l15k4** [comment]
@tgrabiec perfect, I will use : 
http://www.scylladb.com/2017/04/04/counters/
to track number of updates and keep that Set size "sane" without an extra size-query checking overhead.

I guess the counters are part of the new AMI for i3 instances http://www.scylladb.com/2017/05/10/faster-and-better-what-to-expect-running-scylla-on-aws-i3-instances/

**Comment by tgrabiec** [comment]
2017-05-24 19:44 GMT+02:00 Jakub Liska <notifications@github.com>:

> @tgrabiec <https://github.com/tgrabiec> perfect, I will use :
> http://www.scylladb.com/2017/04/04/counters/
> to track number of updates and keep that Set size "sane" without an extra
> size-query checking overhead.
>
What will you do if such counter indicates that the size of the set is too
large?

**Comment by l15k4** [comment]
@tgrabiec 
1) I would change this : 
```
CREATE TABLE mid (
    id text,
    pid text,
    gwid_set set<text>,
    diid_set set<text>,
    PRIMARY KEY (id,pid)
);
```
to 
```
CREATE TABLE mid (
    id text,
    pid text,
    type text,
    uid_set set<text>,
    PRIMARY KEY (id,pid,type)
);
```
which should improve the cell size problem, right?

2) using Counters, I would keep track of number of `mid` updates and when the number of updates is divisible by 10 000, I would check the number of actual `uids` and if it is > than 5000 I would mark that mid with an extra column as "overflowed" and : 
    a) stop updating this row entirely and start loosing uids
    b) have an extra "dump" table where these leftovers would be indexed to not loose it entirely

**Comment by l15k4** [comment]
@tgrabiec I'm thinking, is scylla `map` implemented the same way as Cassandra `map`? Each key being a dynamic column? This could theoretically remove the size restriction, right? If I used `map` instead of `set`. Wdyt?

**Comment by avikivity** [comment]
map and set are implemented in a very similar way.

Currently, both are limited to about 256k, above which you can encounter problems.

Fairly soon, we will lift that limitation, but it will still be limited to a few megabytes.  Likely Cassandra has a similar limitation.

**Comment by l15k4** [comment]
@avikivity There might be a mathematical solution to this problem, let's say you persist time series data, you could have : 
```
CREATE TABLE foo (
    pk text,
    id text,
    time0 set<timestamp>,
    time1 set<timestamp>,
    time2 set<timestamp>,
    time3 set<timestamp>,
    time4 set<timestamp>,
    time5 set<timestamp>,
    time6 set<timestamp>,
    time7 set<timestamp>,
    time8 set<timestamp>,
    time9 set<timestamp>,
    timex set<timestamp>,
    PRIMARY KEY (pk,id)
);
```
And then calculate where to store the timestamp this way `millisSinceEpoch % x` ... Retrieving all timestamps would be a matter of merging those sets on client side ...

This might work, right? To circumvent the collection size problem ...

**Comment by bhalevy** [comment]
This issue aged out.
There were MANY changes around error handling on the sstable writing and file close path.
@tgrabiec is there still anything specific that needs to be done to fix and close this issue?

**Comment by mykaul** [comment]
> This issue aged out. There were MANY changes around error handling on the sstable writing and file close path. @tgrabiec is there still anything specific that needs to be done to fix and close this issue?

I'm closing it, let's open a clean one if it's still relevant (I think the discussion was also hijacked in the middle to a different issue, making it even more complex to understand...)
